### PR TITLE
[Enhancement] extract the correct body if s3 notify via SNS->SQS

### PR DESCRIPTION
**Pull request by christianherweg0807**
## What does this PR do?
This PR parses the SQS Message Body and tries to match to a SNS Message Structure. If BodyJSON.TopicArn is not empty, it delivers  the bodyJSON.Message as body instead of m.Body.

## Use cases
Configless detections of S3 notifications delivered via SNS to a SQS Queue.

### CODE REVIEW

**Comment by christianherweg0807** [comment]
Now we´ve added the tests. Ready for review. Sorry for the delay.

**Review by kaiyan-sheng — COMMENTED** [review]
Thank you for working on this. Could you add a  changelog entry for this fix please?

**Comment by christianherweg0807** [comment]
Added a simple line to the changelog. Any other documentation needed?

**Comment by christianherweg0807** [comment]
Anything, I could do here? @kaiyan-sheng

File: x-pack/filebeat/input/awss3/collector.go [hunk 1]
```diff
@@ -245,13 +245,30 @@ func getRegionFromQueueURL(queueURL string, endpoint string) (string, error) {
...
-		} else {
-			return nil, fmt.Errorf("json unmarshal sqs message body failed: %w", err)
+	var bodyJSON struct {
+		TopicArn          string
+		Message           string
+		MessageAttributes map[string]struct{ Value string }
+	}
+	if err := json.Unmarshal([]byte(*m.Body), &bodyJSON); err == nil && bodyJSON.TopicArn != "" {
+		err := json.Unmarshal([]byte(bodyJSON.Message), &msg)
+		if err != nil {
```
Comment (kaiyan-sheng) [inline x-pack/filebeat/input/awss3/collector.go:255]: Seems like these two error checks have duplicate code, could you simplify this part please?

File: x-pack/filebeat/input/awss3/collector.go [hunk 2]
```diff
@@ -245,13 +245,30 @@ func getRegionFromQueueURL(queueURL string, endpoint string) (string, error) {
...
-		if jsonError, ok := err.(*json.SyntaxError); ok {
-			return nil, fmt.Errorf("json unmarshal sqs message body failed at offset %d with syntax error: %w", jsonError.Offset, err)
-		} else {
-			return nil, fmt.Errorf("json unmarshal sqs message body failed: %w", err)
+	var bodyJSON struct {
+		TopicArn          string
+		Message           string
+		MessageAttributes map[string]struct{ Value string }
+	}
+	if err := json.Unmarshal([]byte(*m.Body), &bodyJSON); err == nil && bodyJSON.TopicArn != "" {
```
Comment (kaiyan-sheng) [inline x-pack/filebeat/input/awss3/collector.go:253]: Could you add a comment here why it's checking bodyJSON.TopicArn please?

**Comment by kaiyan-sheng** [comment]
@christianherweg0807 Sorry for the delay on this!! Thank you for your contribution. With this change, does that mean users can setup S3 bucket to create and subscribe to an SNS topic?

**Comment by christianherweg0807** [comment]
Before this change filebeat only supports this setup:
S3SQS
If the SQS Message was delivered via S3->SNS there is an additional layer in the SQS messsage. With this PR Filebeat detect´s  this layer. Now beats supports this, too:
Ss3SNSSQS

regards
Christian

**Comment by kaiyan-sheng** [comment]
Thank you @christianherweg0807 so much for the information! I think this is a great addition to s3 input, maybe we should also add more in the [documentation](https://github.com/elastic/beats/blob/master/x-pack/filebeat/docs/inputs/input-aws-s3.asciidoc )?

File: x-pack/filebeat/input/awss3/collector.go (hunk 2 above)
Comment (christianherweg0807) [reply x-pack/filebeat/input/awss3/collector.go:256]: I´ve added a code comment. See:
https://docs.aws.amazon.com/sns/latest/dg/sns-sqs-as-subscriber.html

File: x-pack/filebeat/input/awss3/collector.go (hunk 1 above)
Comment (christianherweg0807) [reply x-pack/filebeat/input/awss3/collector.go]: Deduplicated this. You are right...

File: x-pack/filebeat/input/awss3/collector.go (hunk 2 above)
Comment (kaiyan-sheng) [reply x-pack/filebeat/input/awss3/collector.go:282]: Thanks!

**Comment by kaiyan-sheng** [comment]
seems like CI is not happy, maybe you need `go mod tidy`?

**Comment by christianherweg0807** [comment]
Sorry, i missed a variable scope...

**Comment by kaiyan-sheng** [comment]
@christianherweg0807 Hmm seems like CI is not happy with go.sum? 
```
[2021-07-27T05:00:16.151Z] + make check-no-changes

[2021-07-27T05:00:18.060Z] diff --git a/go.sum b/go.sum

[2021-07-27T05:00:18.060Z] index a2a089d0d1..19b2ae2161 100644

[2021-07-27T05:00:18.060Z] --- a/go.sum

[2021-07-27T05:00:18.060Z] +++ b/go.sum

[2021-07-27T05:00:18.060Z] @@ -700,7 +700,6 @@ github.com/stretchr/testify v1.5.1/go.mod h1:5W2xD1RspED5o8YsWQXVCued0rvSQ+mT+I5

[2021-07-27T05:00:18.060Z]  github.com/stretchr/testify v1.6.1/go.mod h1:6Fq8oRcR53rry900zMqJjRRixrwX3KX962/h/Wwjteg=

[2021-07-27T05:00:18.060Z]  github.com/stretchr/testify v1.7.0 h1:nwc3DEeHmmLAfoZucVR881uASk0Mfjw8xYJ99tb5CcY=

[2021-07-27T05:00:18.060Z]  github.com/stretchr/testify v1.7.0/go.mod h1:6Fq8oRcR53rry900zMqJjRRixrwX3KX962/h/Wwjteg=

[2021-07-27T05:00:18.060Z] -github.com/syndtr/gocapability v0.0.0-20170704070218-db04d3cc01c8 h1:zLV6q4e8Jv9EHjNg/iHfzwDkCve6Ua5jCygptrtXHvI=

[2021-07-27T05:00:18.060Z]  github.com/syndtr/gocapability v0.0.0-20170704070218-db04d3cc01c8/go.mod h1:hkRG7XYTFWNJGYcbNJQlaLq0fg1yr4J4t/NcTQtrfww=

[2021-07-27T05:00:18.060Z]  github.com/tsg/go-daemon v0.0.0-20200207173439-e704b93fd89b h1:X/8hkb4rQq3+QuOxpJK7gWmAXmZucF0EI1s1BfBLq6U=

[2021-07-27T05:00:18.060Z]  github.com/tsg/go-daemon v0.0.0-20200207173439-e704b93fd89b/go.mod h1:jAqhj/JBVC1PwcLTWd6rjQyGyItxxrhpiBl8LSuAGmw=

[2021-07-27T05:00:19.439Z] go.sum: needs update

[2021-07-27T05:00:19.439Z] Makefile:113: recipe for target 'check-no-changes' failed

[2021-07-27T05:00:19.439Z] make: *** [check-no-changes] Error 1

script returned exit code 2
```

**Comment by kaiyan-sheng** [comment]
@christianherweg0807 Hi sorry you have to deal with the conflict because there were two big prs got merged for s3 input.

**Comment by kaiyan-sheng** [comment]
@christianherweg0807 Thank you so much for your initiative on this!! I'm closing this PR and will use https://github.com/elastic/beats/pull/28800 to continue to work.

**Pull request closed without merging**